lib/reline/line_editor.rb

```diff
@@ -831,27 +831,23 @@ def add_dialog_proc(name, p, context = nil)
       dialog.column = 0
       dialog.width = @screen_size.last
     end
+    face = Reline::Face[dialog_render_info.face || :default]
+    scrollbar_sgr = face.scrollbar
+    default_sgr = face.default
     dialog.contents = contents.map.with_index do |item, i|
-      if i == pointer
-        fg_color = dialog_render_info.pointer_fg_color
-        bg_color = dialog_render_info.pointer_bg_color
-      else
-        fg_color = dialog_render_info.fg_color
-        bg_color = dialog_render_info.bg_color
-      end
+      line_sgr = i == pointer ? face.enhanced : default_sgr
```

Comment:
This will cause error when users override `completion_dialog` without defining `:enhanced`.


```rb
# .irbrc
Reline::Face.config(:completion_dialog) do |face|
  face.define :default, foreground: :white, background: :cyan
  face.define :scrollbar, foreground: :white, background: :cyan
end
```

And it's also why I think the method definition approach is not good for maintenance as we'll need to use `face,respond_to?(:enhanced)` to safeguard it, which will make it very hard to refactor in the future as we won't know if the method is not defined due to implementation change or value not being set correctly.

Reply:
> Can we also have some yamatanooroti tests for this change? One for testing the default completion_dialog setting works end-to-end, and others to test if different kinds of user-provided settings could cause issues to the dialog.

I'm reluctant to do that. reason:
- Reline's rendering tests fork another process with `multiline_repl` command, so applying arbitrary Reline::Face settings only to tests related to dialogs looks difficult or requires major modification
- It may be possible to repeat the entire test case of test_yamatanooroti under two different Reline::Face settings, but test_yamatanooroti is very time-consuming. It doesn't seem worth doubling your time
- Originally, test_yamatanooroti was not interested in text decoration. Escape sequences that change the color of text are simply ignored and have no effect on the rendering result. I think it is sufficient that Reline::Face::SGR_PARAMETERS is private_constant and immutable

BTW, I believe that the maintenance of yamatanooroti is difficult, and I would like to propose an alternative or an improved testing method, that is a more effective one, in the future.

Reply:
> This will cause error when users override completion_dialog without defining :enhanced.

I fixed it. 

> And it's also why I think the method definition approach is not the best for maintenance as we'll need to use face.respond_to?(:enhanced) to safeguard it. It will then be hard to refactor in the future as we won't know if the method is not defined due to implementation change or value not being set correctly.

OK, I changed it into hash-accessor style.
It will be slower a bit though, I'm fine.

And I added an accessor to list the config:

```ruby
Reline::Face.configs
=>
{:default=>
  {:default=>{:style=>:default, :escape_sequence=>"\e[0m"},
   :enhanced=>{:style=>:default, :escape_sequence=>"\e[0m"},
   :scrollbar=>{:style=>:default, :escape_sequence=>"\e[0m"}},
 :completion_dialog=>
  {:default=>{:foreground=>:white, :background=>:cyan, :escape_sequence=>"\e[0m\e[37;46m"},
   :enhanced=>{:foreground=>:white, :background=>:magenta, :escape_sequence=>"\e[0m\e[37;45m"},
   :scrollbar=>{:foreground=>:white, :background=>:cyan, :escape_sequence=>"\e[0m\e[37;46m"}}}
```

These two expressions return the same value:

```
Reline::Face[:completion_dialog][:enhanced]
=> "\e[0m\e[37;45m"
Reline::Face.configs[:completion_dialog][:enhanced][:escape_sequence]
=> "\e[0m\e[37;45m"
```

However, the latter is disadvantageous in terms of performance due to the large number of method calls, so it is intended to be used only for debugging.

Reply:
Do you have benchmarks about the performance? `Reline` already performs a large number of method calls on each dialog rendering call (e.g. when `render_dialog_changes` is called). So while I'm aware of the increased number of method calls, I'm curious to see how it'd impact the overall performance. But at the moment, I'm not too concerned about the overhead this change would bring compare to other calculations `Reline` performs.

Reply:
Probably, even if Ruby has a lot of internal processing, there are many cases where the I/O wait time is greater, and humans cannot perceive a delay of about 10ms, so it would not be a big problem.
As long as we don't get complaints from users that it's slow, we shouldn't worry about it too much.
At least in my environment, I don't feel that rendering is slow.